### PR TITLE
Implement nodeUniqueIndexSeek in Kernel API

### PR DESCRIPTION
**Pull request by pontusmelke**
Builds on top of https://github.com/neo4j/neo4j/pull/11243

### CODE REVIEW

File: community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/Read.java [hunk 1]
```diff
@@ -41,6 +39,20 @@
...
+     * Note that this is a very special method and should be use with caution. It has special locking semantics in
+     * order to facilitate unique creation of nodes. If a node is found; a shared lock for the index entry will be
+     * held whereas if no node is found we will hold onto an exclusive lock until the close of the transaction.
+     *
+     * @param index {@link IndexReference} referencing index to query.
+     * @param indexOrder requested {@link IndexOrder} of result. Must be among the capabilities of
+     * {@link IndexReference referenced index}, or {@link IndexOrder#NONE}.
+     * @param predicates Combination of {@link IndexQuery.ExactPredicate index queries} to run against referenced index.
+     */
+    long nodeUniqueIndexSeek( IndexReference index, IndexOrder indexOrder, IndexQuery.ExactPredicate... predicates )
```
Comment (fickludd) [inline community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/Read.java]: Hum hum, how about omitting `IndexOrder` here, as the order of a single node doesn't seem that interesting?

File: community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/TransactionBoundQueryContext.scala [hunk 2]
```diff
@@ -814,8 +813,7 @@ sealed class TransactionBoundQueryContext(val transactionalContext: Transactiona
 
   override def dropIndexRule(descriptor: IndexDescriptor): Unit =
     transactionalContext.kernelTransaction.schemaWrite().indexDrop(
-      new DefaultCapableIndexReference(false, IndexCapability.NO_CAPABILITY, null, descriptor.label,
-                                       descriptor.properties.map(_.id):_*)
+      DefaultIndexReference.general(descriptor.label, descriptor.properties.map(_.id):_*)
```
Comment (fickludd) [inline community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/TransactionBoundQueryContext.scala:816]: Yeah, this becomes much nicer for the API users. Like it.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/Operations.java [hunk 3]
```diff
@@ -92,12 +93,12 @@
 import static java.lang.Math.max;
 import static java.lang.Math.min;
 import static org.neo4j.internal.kernel.api.exceptions.schema.ConstraintValidationException.Phase.VALIDATION;
-import static org.neo4j.internal.kernel.api.exceptions.schema.SchemaKernelException.OperationContext.CONSTRAINT_CREATION;
+import static org.neo4j.internal.kernel.api.exceptions.schema.SchemaKernelException.OperationContext
+        .CONSTRAINT_CREATION;
```
Comment (fickludd) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/Operations.java]: whoops?

**Review by fickludd — CHANGES_REQUESTED** [review]
I think we should remove `IndexOrder` from `nodeUniqueIndexSeek`, otherwise looks good. I think the code becomes cleaner and easier to read in many cases when changing test. It's a good sign.

File: community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/Read.java (hunk 1 above)
Comment (pontusmelke) [reply community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/Read.java]: you and your logic

**Review by fickludd — APPROVED** [review]

**Pull request merged**